### PR TITLE
Network update

### PR DESCRIPTION
**Pull request by mfornet**
Implement several points from #1363 
- [x] Round robin to send messages.
- [x] Recalculate routing table after small period of time to allow implicit batch computation.
- [x] Edge removal and automatic nonce updates.
- [x] Tests that covers edge removal.

### CODE REVIEW

File: chain/network/Cargo.toml [hunk 1]
```diff
@@ -17,6 +17,7 @@ rand = "0.6.5"
 reed-solomon-erasure = "3.1.1"
 byteorder = "1.2"
 lazy_static = "1.4"
+itertools = "0.8.0"
```
Comment (bowenwang1996) [inline chain/network/Cargo.toml]: We intentionally avoid using this to reduce dependencies. Unless there is a strong reason we should not use it.

Comment (mfornet) [reply chain/network/Cargo.toml]: Ohh, sad to hear that, this allowed me to make some idiomatic piece of code, which is cleaner with itertools than without it, but of course it is not mandatory. I can do it by myself.

File: chain/network/src/routing.rs [hunk 2]
```diff
@@ -224,21 +262,61 @@ impl RoutingTable {
...
+            // nonce is greater than some threshold increase the lowest nonce to be at least
+            // max nonce - threshold.
+
+            let (min_v, max_v) = routes
+                .iter()
+                .map(|peer_id| {
+                    let nonce = self.route_nonce.get(&peer_id).cloned().unwrap_or(0usize);
+                    (nonce, peer_id.clone())
+                })
+                .fold((None, None), |(min_v, max_v), current| {
```
Comment (bowenwang1996) [inline chain/network/src/routing.rs]: Why not fold directly?

Comment (mfornet) [reply chain/network/src/routing.rs]: No need to, that was an old piece of code refactorized. I'll remove `map`.

File: chain/network/src/peer_manager.rs [hunk 3]
```diff
@@ -179,28 +188,40 @@ impl PeerManagerActor {
         });
     }
 
-    fn remove_active_peer(&mut self, peer_id: &PeerId) {
-        // TODO(MarX, #1312): Trigger actions after a peer is removed successfully regarding networking
-        //  (Write specification about the actions)
+    fn remove_active_peer(&mut self, ctx: &mut Context<Self>, peer_id: &PeerId) {
+        println!("\n REMOVE ACTIVE PEER me:{:?} other:{:?}\n", self.peer_id, peer_id);
```
Comment (ilblackdragon) [inline chain/network/src/peer_manager.rs]: remove print. you can use debug! if you want

File: chain/network/src/peer_manager.rs [hunk 4]
```diff
@@ -42,6 +42,12 @@ use crate::NetworkClientResponses;
 
 /// How often to request peers from active peers.
 const REQUEST_PEERS_SECS: i64 = 60;
+/// How much time to wait (in milliseconds) after we send update nonce request before disconnecting.
+/// This number should be large to handle pair of nodes with high latency.
+const WAIT_ON_TRY_UPDATE_NONCE: u64 = 6000;
```
Comment (ilblackdragon) [inline chain/network/src/peer_manager.rs]: it's usually preferred to use 6_000

File: chain/network/src/peer_manager.rs [hunk 5]
```diff
@@ -320,9 +406,10 @@ impl PeerManagerActor {
                 .and_then(move |res, act, _| {
                     if res.is_abusive {
                         trace!(target: "network", "Banning peer {} for abuse ({} sent, {} recv)", peer_id1, res.message_counts.0, res.message_counts.1);
-                        // TODO(MarX): I think banning peer here leaves a hanging Peer instance.
-                        //  We should send signal to peer instead, and expect signal back.
-                        // act.ban_peer(&peer_id1, ReasonForBan::Abusive);
+                        // Send ban signal to peer instance. It should send ban signal back and stop the instance.
```
Comment (ilblackdragon) [inline chain/network/src/peer_manager.rs]: TODO for this commented code?

File: chain/network/tests/routing.rs [hunk 6]
```diff
@@ -148,6 +150,7 @@ impl StateMachine {
                             .map_err(|_| ())
                             .and_then(move |res| {
                                 if let NetworkResponses::RoutingTableInfo(routing_table) = res {
+                                    println!("\nROUTING TABLE INFO: {:?}\n", routing_table);
```
Comment (ilblackdragon) [inline chain/network/tests/routing.rs]: remove print here.

**Comment by mfornet** [comment]
@ilblackdragon failing test (`run_nodes_4_4_2`) in my machine takes from 13 to 19 seconds and it only has 20 seconds (that is also the case in `staging`), I think it is too tight for CI and it's failing for no good reason.

**Pull request merged**